daft/execution/physical_plan.py

```diff
@@ -386,40 +390,62 @@ def hash_join(
-        elif len(left_requests) == len(right_requests):
-            # Both plans have progressed equally; alternate between the two plans to avoid starving either one
-            next_plan, next_requests = (left_plan, left_requests) if yield_left else (right_plan, right_requests)
-            yield_left = not yield_left
+    join_tasks: dict[int, SingleOutputPartitionTask[PartitionT]] = {}
+    right_partition_counter = 0
+    while True:
+        # Find all partitions that are ready to be joined
+        ready_partitions = [
+            partition_num
+            for partition_num in left_tasks.keys() & right_tasks.keys()  # Intersection of keys
+            if left_tasks[partition_num].done() and right_tasks[partition_num].done()
+        ]
 
-        try:
-            step = next(next_plan)
-            if isinstance(step, PartitionTaskBuilder):
-                step = step.finalize_partition_task_single_output(stage_id=stage_id)
-                next_requests.append(step)
-            yield step
+        if len(ready_partitions) > 0:
+            # Process all ready pairs
+            for partition in ready_partitions:
+                left_task = left_tasks.pop(partition)
+                right_task = right_tasks.pop(partition)
+                join_task = create_join_step(left_task, right_task)
+                join_tasks[partition] = join_task
+                yield join_task
+        else:
+            try:
+                # Process next right plan step
+                step = next(right_plan)
+                if isinstance(step, PartitionTaskBuilder):
+                    step = step.finalize_partition_task_single_output(stage_id=right_stage_id)
+                    right_tasks[right_partition_counter] = step
+                    right_partition_counter += 1
+                yield step
 
-        except StopIteration:
-            # Left and right child plans have completed.
-            # Are we still waiting for materializations to complete? (We will emit more joins from them).
-            if len(left_requests) + len(right_requests) > 0:
-                logger.debug(
-                    "join blocked on completion of sources.\n Left sources: %s\nRight sources: %s",
-                    left_requests,
-                    right_requests,
-                )
-                yield None
+            except StopIteration:
+                if left_tasks or right_tasks:
+                    logger.debug(
+                        "join blocked on completion of sources.\n Left sources: %s\nRight sources: %s",
+                        left_tasks,
+                        right_tasks,
+                    )
+                    yield None
+                else:
+                    break
 
-            # Otherwise, we are entirely done.
-            else:
-                return
+    # yield results of join tasks in order of partition number
+    for partition in range(len(join_tasks)):
```

Comment:
you don't have to wait until everything is done to do this.

We can put this check in the main while loop and check `join_tasks[last_emmited_idx + 1].done()` and then yield that upsteam. This allows us to emit completed joins while other partitions are still computing.